File: code/R/ch10/10.03.R
## 10.3 タイタニック

library(caret)
library(PRROC)
library(tidyverse)

my_url <- str_c("https://raw.githubusercontent.com/taroyabuki",
                "/fromzero/master/data/titanic.csv")
my_data <- read_csv(my_url)

head(my_data)
#> # A tibble: 6 x 4
#>   Class Sex   Age   Survived
#>   <chr> <chr> <chr> <chr>
#> 1 1st   Male  Child Yes
#> 2 1st   Male  Child Yes
#> 3 1st   Male  Child Yes
#> 4 1st   Male  Child Yes
#> 5 1st   Male  Child Yes
#> 6 1st   Male  Adult No

### 10.3.1 質的入力変数の扱い方

### 10.3.2 決定木の訓練

my_model <- train(form = Survived ~ ., data = my_data, method = "rpart2",
                  tuneGrid = data.frame(maxdepth = 2),
                  trControl = trainControl(method = "LOOCV"))

### 10.3.3 決定木の描画

rpart.plot::rpart.plot(my_model$finalModel, extra = 1)

### 10.3.4 決定木の評価

my_model$results
#>   maxdepth  Accuracy     Kappa
#> 1        2 0.7832803 0.4096365

y <- my_data$Survived
tmp <- my_model %>% predict(newdata = my_data, type = "prob")
y_score <- tmp$Yes

my_roc <- roc.curve(scores.class0 = y_score[y == "Yes"],
                    scores.class1 = y_score[y == "No"],
                    curve = TRUE)
my_roc$auc
#> [1] 0.7114887

my_roc %>% plot(xlab = "False Positive Rate",
                ylab = "True Positive Rate",
                legend = FALSE)

### 10.3.5 補足：質的入力変数の扱い

X <- my_data %>% select(Class) # 質的入力変数
y <- my_data$Survived          # 出力変数

options(warn = -1) # これ以降，警告を表示しない．
my_model1 <- train(x = X, y = y, method = "rpart2",
                   tuneGrid = data.frame(maxdepth = 2),
                   trControl = trainControl(method = "LOOCV"))
options(warn = 0)  # これ以降，警告を表示する．

rpart.plot::rpart.plot(my_model1$finalModel, extra = 1)
my_model1$results
#>   maxdepth  Accuracy     Kappa
#> 1        2 0.7137665 0.2373133

my_data2 <- my_data %>%
  dummyVars(formula = Survived ~ Class) %>%
  predict(my_data) %>%
  as.data.frame %>%
  mutate(Survived = my_data$Survived)

my_model2 <- train(form = Survived ~ ., data = my_data2, method = "rpart2",
                   tuneGrid = data.frame(maxdepth = 2),
                   trControl = trainControl(method = "LOOCV"))
rpart.plot::rpart.plot(my_model2$finalModel, extra = 1)
my_model2$results
#>   maxdepth  Accuracy     Kappa
#> 1        2 0.7137665 0.2373133

my_model3 <- train(form = Survived ~ Class, data = my_data, method = "rpart2",
                   tuneGrid = data.frame(maxdepth = 2),
                   trControl = trainControl(method = "LOOCV"))
rpart.plot::rpart.plot(my_model3$finalModel, extra = 1)
my_model3$results
#>   maxdepth  Accuracy     Kappa
#> 1        2 0.6915039 0.2674485

File: code/py/ch10/10.03.py
## 10.3 タイタニック

import graphviz
import pandas as pd
from sklearn import tree
from sklearn.metrics import roc_curve, RocCurveDisplay, auc
from sklearn.model_selection import cross_val_score, LeaveOneOut
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder

my_url = ('https://raw.githubusercontent.com/taroyabuki'
          '/fromzero/master/data/titanic.csv')
my_data = pd.read_csv(my_url)

my_data.head()
#>   Class   Sex    Age Survived
#> 0   1st  Male  Child      Yes
#> 1   1st  Male  Child      Yes
#> 2   1st  Male  Child      Yes
#> 3   1st  Male  Child      Yes
#> 4   1st  Male  Child      Yes

### 10.3.1 質的入力変数の扱い方

### 10.3.2 決定木の訓練

X, y = my_data.iloc[:, 0:3], my_data.Survived

my_pipeline = Pipeline([
    ('ohe', OneHotEncoder(drop='first')),
    ('tree', tree.DecisionTreeClassifier(max_depth=2, random_state=0,
                                         min_impurity_decrease=0.01))])
my_pipeline.fit(X, y)

### 10.3.3 決定木の描画

my_enc  = my_pipeline.named_steps['ohe']  # パイプラインからエンコーダを取り出す．
my_tree = my_pipeline.named_steps['tree'] # パイプラインから木を取り出す．

my_dot = tree.export_graphviz(
    decision_tree=my_tree,
    out_file=None,
    feature_names=my_enc.get_feature_names_out() if hasattr(my_enc, 'get_feature_names_out') else my_enc.get_feature_names(),
    class_names=my_pipeline.classes_,
    filled=True)
graphviz.Source(my_dot)

### 10.3.4 決定木の評価

my_scores = cross_val_score(
    my_pipeline, X, y,
    cv=LeaveOneOut(),
    n_jobs=-1)
my_scores.mean()
#> 0.7832803271240345

tmp = pd.DataFrame(
    my_pipeline.predict_proba(X),
    columns=my_pipeline.classes_)
y_score = tmp.Yes

my_fpr, my_tpr, _ = roc_curve(y_true=y,
                              y_score=y_score,
                              pos_label='Yes')
my_auc = auc(x=my_fpr, y=my_tpr)
my_auc
#> 0.7114886868858494

RocCurveDisplay(fpr=my_fpr, tpr=my_tpr, roc_auc=my_auc).plot()

### 10.3.5 補足：質的入力変数の扱い
